blueprint/core/blueprint_EcmascriptEngine.cpp

```diff
@@ -252,8 +252,101 @@ namespace blueprint
             return result;
         }
 
+        struct TimeoutFunctionManager : private juce::MultiTimer
+        {
+            ~TimeoutFunctionManager() override {
+                for(const auto &[id, timer] : timeoutFunctions)
+                    stopTimer(id);
+            }
+
+            juce::var clearTimeout(const int id)
+            {
+                stopTimer(id);
+                const auto f = timeoutFunctions.find(id);
+                if(f != timeoutFunctions.cend())
+                    timeoutFunctions.erase(f);
+                return juce::var();
+            }
+
+            int newTimeout(const juce::var::NativeFunction f, const int timeoutMillis, const std::vector<juce::var>&& args, const bool repeats=false)
+            {
+                static int nextId = 0;
+                timeoutFunctions.emplace(nextId, TimeoutFunction(f, std::move(args), repeats));
+                startTimer(nextId, timeoutMillis);
+                return nextId++;
+            }
+
+            void timerCallback(int id) override
+            {
+                const auto f = timeoutFunctions.find(id);
+                if(f != timeoutFunctions.cend())
+                {
+                    const auto cb = f->second;
+                    std::invoke(cb.f, juce::var::NativeFunctionArgs(juce::var(), cb.args.data(), static_cast<int>(cb.args.size())));
+                    if(!cb.repeats)
+                    {
+                        stopTimer(id);
+                        timeoutFunctions.erase(f);
+                    }
+                }
+            }
+
+            private:
+                struct TimeoutFunction
+                {
+                    TimeoutFunction(const juce::var::NativeFunction _f, const std::vector<juce::var> &&_args, const bool _repeats=false)
+                    : f(_f), args(std::move(_args)), repeats(_repeats) {}
+
+                    const juce::var::NativeFunction f;
+                    std::vector<juce::var> args;
+                    const bool repeats;
+                };
+
+                std::map<int, TimeoutFunction> timeoutFunctions;
+        };
+
+        // IsSetter is true for setTimeout / setInterval
+        // and false for clearTimeout / clearInterval
+        template <bool IsSetter = false, bool Repeats = false, typename MethodType>
+        void registerNativeTimerFunction(const char* name, MethodType method)
+        {
+            registerNativeProperty(name, juce::var::NativeFunction([this, name, method] (const juce::var::NativeFunctionArgs& _args) -> juce::var {
+                if constexpr (IsSetter)
+                {
+                    if(_args.numArguments < 2 || !_args.arguments[0].isMethod() || !_args.arguments[1].isDouble())
+                        throw Error(juce::String(name) + " requires a callback and time in milliseconds");
```

Comment:
This gets caught and raised into the js env via the code you have below, right?

Reply:
It does, but the `duk_throw` that happens down there will only result in a red box when called from one of the `ReactApplicationRoot` methods.  If a bad call is made from within a timer callback, we get termination.

e.g.
```
setTimeout('nonsense'); // red box 👍🏾 
setTimeout(()=>setInterval('nonsense'), 1000); // termination 👎🏾 
```

So that's our dilemma, in the catch code inside `LambdaHelper` we have to decide not to re-throw to prevent this kind of termination, or to have a mechanism of pushing the `Error` to some kind of queue that `ReactApplicationRoot` is servicing (and handling with red-boxes), or stick with this possibility of termination.  I reckon somebody around here will have a much better idea than all of the above.

Reply:
Hmm interesting. Yea, treating this kind of error flow properly I think will be a primary focus after hitting our beta milestone. Let's not worry about it too much here, no need to make 1 issue into 2 :)

Reply:
Cool, I will certainly open an issue if this gets merged.